jax/scipy/sparse/linalg.py

```diff
@@ -143,34 +231,67 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
   scipy.sparse.linalg.cg
   jax.lax.custom_linear_solve
   """
-  if x0 is None:
-    x0 = tree_map(jnp.zeros_like, b)
 
-  b, x0 = device_put((b, x0))
+  return isolve(_cg_solve,
+                A=A, b=b, x0=x0, tol=tol, atol=atol,
+                maxiter=maxiter, M=M)
 
-  if maxiter is None:
-    size = sum(bi.size for bi in tree_leaves(b))
-    maxiter = 10 * size  # copied from scipy
 
-  if M is None:
-    M = _identity
+def bicgstab(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
+  """Use Bi-Conjugate Gradient Stable iteration to solve ``Ax = b``.
 
-  if tree_structure(x0) != tree_structure(b):
-    raise ValueError(
-        'x0 and b must have matching tree structure: '
-        f'{tree_structure(x0)} vs {tree_structure(b)}')
+  The numerics of JAX's ``bicgstab`` should exact match SciPy's
+  ``bicgstab`` (up to numerical precision), but note that the interface
+  is slightly different: you need to supply the linear operator ``A`` as
+  a function instead of a sparse matrix or ``LinearOperator``.
 
-  if _shapes(x0) != _shapes(b):
-    raise ValueError(
-        'arrays in x0 and b must have matching shapes: '
-        f'{_shapes(x0)} vs {_shapes(b)}')
+  Derivatives of ``bicgstab`` are implemented via implicit
+  differentiation with another ``bicgstab`` solve, rather than by
+  differentiating *through* the solver. They will be accurate only if
+  both solves converge.
 
-  cg_solve = partial(
-      _cg_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
-  # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
-  symmetric = all(map(real_valued, tree_leaves(b)))
-  x = lax.custom_linear_solve(
-      A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
-  info = None  # TODO(shoyer): return the real iteration count here
-  return x, info
+  Parameters
+  ----------
+  A : function
+      Function that calculates the matrix-vector product ``Ax`` when called
+      like ``A(x)``. ``A`` must represent a hermitian, positive definite
```

Comment:
I thought bicgstab works on arbitrary matrices, without requiring symmetry? It would be good to clarify this :)

Reply:
That's a copying error, you are correct!